rfcs/77-dev-tools-abstraction.md

```diff
@@ -0,0 +1,658 @@
+
+```rust
+/// Sets the player's gold to the provided value.
+#[derive(Reflect, FromReflect, Debug)]
+struct SetGold {
+    amount: u64,
+}
+
+impl Command for SetGold {
+    fn apply(&self, world: &mut World){
+        let mut current_gold = world.resource_mut::<Gold>();
+        current_gold.0 = self.amount;
+    }
+}
+
+impl DevCommand for SetGold {}
+
+impl FromStr for SetGold {
+    fn from_str(s: &str) -> Result<Self, DevToolParseError>{
+        let parts = s.split_whitespace();
+        let name = parts.iter.next()?;
+        if name != self.name() {
+            return Err(DevToolParseError::InvalidName);
+        }
+
+        let amount_string = parts.iter.next()?;
+        let amount = amount_string.parse()?;
+
+        Ok(SetGold {amount} )
+    }
+}
+```
+
+### Conventions for building dev tools
+
+Not everything can or should be defined by a trait! To supplement the `ModalDevTool` trait, we recommend that Bevy and its ecosystem follow the following conventions:
+
+1. Dev tools can be toggled at compile time.
+2. Dev tools can be toggled at run time.
+3. Dev tools implement the `ModalDevTool` trait, and if the corresponding feature is enabled, are added to the app via `.init_dev_tool` or `.insert_dev_tool`.
+   1. This can be done either in the main plugin or a dedicated dev tools plugin.
+   1. If configuration is required, splitting this into a dedicated dev tools plugin is preferred.
+4. Dev tools are disabled by default, both at compile and run-time.
+5. Each dev tools should be configured independently via a single resource.
+6. If a meaningful defaults exist for fields on dev commands, that field should be modelled as an `Option`.
+   1. This allows CLI-style toolboxes to more easily populate your struct.
+
+### Building toolboxes
+
+Let's take a look at how this all fits together by attempting to build a Quake-style dev console, much like [`bevy-console`](https://github.com/RichoDemus/bevy-console).
+Obviously, this is too big to cover properly in an RFC example, so we're going to hand-wave the user input side of things by showing how you might deal with the parsed events, and simplify output by just logging to `println!`.
+
+Our first task is getting the list of dev tools.
+
+```rust
+#[derive(Event)]
+struct ListDevTools;
+
+fn list_dev_tools(mut event_reader: EventReader<ListDevTools>, dev_tools_registry: Res<DevToolsRegistry>){
+    // Clear the events, and act if at least one is found
+    if event_reader.drain().next().is_some() {
+        println!("Available modal dev tools:");
+        for (_component_id, modal_dev_tool) in dev_tools_registry.iter_modal_dev_tools() {
+            println!("{}", modal_dev_tool.type_name());
+            println!("{}", modal_dev_tool.docs());
+
+            for field in modal_dev_tools.fields(){
+                println!("{}", field.type_name());
+                println!("{}", field.docs())
```

Comment:
I feel like this method needs to be added to traits, otherwise where does it come from 
Perhaps this is a plan to add a short description to the tools?

Reply:
I remember being able to do this with  reflection: we need to figure out how to actually access that information still.

Reply:
I just realised that maybe the RFC doesn't explicitly say this (or I missed it), but it is supposed to provide in near future a mechanism for automatic parsing of tool structures from comman line interface based on reflection?

Reply:
In terms of not just providing the FromStr mechanism, but being able to show the fields, their types, description in the CLI, and then from the standard CLI string atomatically fill in the fields of the tool's structure, bypassing FromStr<T>?

Reply:
> I remember being able to do this with reflection: we need to figure out how to actually access that information still.

This method does exist, oops, sorry for the incorrect question. Although it is not available by default, but only by non-default flag "documentation" in bevy_reflect crate.

Reply:
Yeah, this was simply pseudo-code: we need to figure out how to actually do this properly still.